tiledb/dimension.py

```diff
@@ -0,0 +1,365 @@
+    elif dtype.kind == "M":
+        info = np.iinfo(np.int64)
+        date_unit = np.datetime_data(dtype)[0]
+        # +1 to exclude NaT
+        dtype_min = np.datetime64(info.min + 1, date_unit)
+        dtype_max = np.datetime64(info.max, date_unit)
+    else:
+        raise TypeError(f"invalid Dim dtype {dtype!r}")
+    return (dtype_min, dtype_max)
+
+
+def _tiledb_cast_tile_extent(tile_extent: Any, dtype: np.dtype) -> np.array:
+    """Given a tile extent value, cast it to np.array of the given numpy dtype."""
+    # Special handling for datetime domains
+    if dtype.kind == "M":
+        date_unit = np.datetime_data(dtype)[0]
+        if isinstance(tile_extent, np.timedelta64):
+            extent_value = int(tile_extent / np.timedelta64(1, date_unit))
+            tile_size_array = np.array(np.int64(extent_value), dtype=np.int64)
+        else:
+            tile_size_array = np.array(tile_extent, dtype=np.int64)
+    else:
+        tile_size_array = np.array(tile_extent, dtype=dtype)
+
+    if tile_size_array.size != 1:
+        raise ValueError("tile extent must be a scalar")
+
+    return tile_size_array
+
+
+def _tiledb_cast_domain(
+    domain, tiledb_dtype: lt.DataType
+) -> Tuple[np.generic, np.generic]:
+    np_dtype = numpy_dtype(tiledb_dtype)
+
+    if tiledb_type_is_datetime(tiledb_dtype):
+        date_unit = np.datetime_data(np_dtype)[0]
+        return (
+            np.datetime64(domain[0], date_unit),
+            np.datetime64(domain[1], date_unit),
+        )
+
+    if tiledb_dtype in (
+        lt.DataType.STRING_ASCII,
+        lt.DataType.STRING_UTF8,
+        lt.DataType.BLOB,
+    ):
+        return domain
+
+    return (np_dtype(domain[0]), np_dtype(domain[1]))
+
+
+class Dim(lt.Dimension):
+    """
+    Represents a TileDB dimension.
+    """
+
+    def __init__(
+        self,
+        name: str = "__dim_0",
+        domain: Tuple[Any, Any] = None,
+        tile: Any = None,
+        filters: Union[FilterList, Sequence[Filter]] = None,
+        dtype: np.dtype = np.uint64,
+        var: bool = None,
+        ctx: "Ctx" = None,
+        _lt_obj: lt.Dimension = None,
+    ):
+        """Class representing a dimension of a TileDB Array.
+
+        :param str name: the dimension name, empty if anonymous
+        :param domain:
+        :type domain: tuple(int, int) or tuple(float, float)
+        :param tile: Tile extent
+        :type tile: int or float
+        :param filters: List of filters to apply
+        :type filters: FilterList
+        :dtype: the Dim numpy dtype object, type object, or string \
+            that can be corerced into a numpy dtype object
+        :raises ValueError: invalid domain or tile extent
+        :raises TypeError: invalid domain, tile extent, or dtype type
+        :raises: :py:exc:`TileDBError`
+        :param tiledb.Ctx ctx: A TileDB Context
+
+        """
+        self._ctx = ctx or default_ctx()
+        _cctx = lt.Context(self._ctx, False)
```

Comment:
Isn't this going to be making a copy?

Reply:
I guess maybe not due to `False`?... But then why do we need a second pybind11 object created at all here? Ideally we should be able to just pass in `self._ctx` on [line 177](https://github.com/TileDB-Inc/TileDB-Py/pull/1327/files#diff-d1da7d0bde5761bf12df07572dd2ee80138056d625c2d9a83f64f066facbb22bR177) rather than unwrapping and re-wrapping it here.